src/System.Text.Primitives/System/Text/Encoding/Utf16/Utf16LittleEndianEncoder.cs

```diff
@@ -2,102 +2,204 @@
+        /// equal to the length of the <paramref name="utf32"/> and <paramref name="charactersWritten"/> will equal the total number of bytes written to
+        /// the <paramref name="utf16"/>.
+        /// 
+        /// On unsuccessful exit, the following conditions can exist.
+        ///  1) If the output buffer has been filled and no more input characters can be encoded, another call to this method with the input sliced to
+        ///     exclude the already encoded characters (using <paramref name="charactersConsumed"/>) and a new output buffer will continue the encoding.
+        ///  2) Encoding may have also stopped because the input buffer contains an invalid sequence.
+        /// </summary>
+        /// <param name="utf32">A span containing a sequence of UTF-32 characters.</param>
+        /// <param name="utf16">A span to write the UTF-16 encoded data into.</param>
+        /// <param name="charactersConsumed">On exit, contains the number of code points that were consumed from the UTF-32 character span.</param>
+        /// <param name="charactersWritten">An output parameter to store the number of bytes written to <paramref name="utf8"/></param>
+        /// <returns>True if the input buffer was fully encoded into the output buffer, otherwise false.</returns>
+        public static bool TryEncode(ReadOnlySpan<uint> utf32, Span<char> utf16, out int charactersConsumed, out int charactersWritten)
         {
-            if (!UnicodeCodePoint.IsSupportedCodePoint(codePoint))
+            charactersConsumed = 0;
+            charactersWritten = 0;
+
+            for (int i = 0; i < utf32.Length; i++)
             {
-                encodedBytes = default(int);
-                return false;
+                int written;
+                if (!TryEncode(utf32[i], utf16, charactersWritten, out written))
+                    return false;
+
+                charactersConsumed++;
+                charactersWritten += written;
             }
 
-            encodedBytes = UnicodeCodePoint.IsBmp(codePoint) ? 2 : 4;
+            return true;
+        }
 
-            if (buffer.Length < encodedBytes)
+        /// <summary>
+        /// Encodes a single UTF-32 character into UTF-16.
+        /// </summary>
+        /// <param name="codePoint">A UTF-32 character to encode.</param>
+        /// <param name="utf16">A span to write the UTF-16 encoded data into.</param>
+        /// <param name="index">An index into the <paramref name="utf16"/> span to to the write operation.</param>
+        /// <param name="charactersWritten">An output parameter to store the number of bytes written to <paramref name="utf16"/></param>
+        /// <returns>True if the input buffer was fully encoded into the output buffer, otherwise false.</returns>
+        public static bool TryEncode(uint codePoint, Span<char> utf16, int index, out int charactersWritten)
+        {
```

Comment:
It is on an internal class, so not really public. It is also being used outside the class by another internal class. I suspect it might go away with optimizations, but we'll have to see.